src/Service/RequestService.php

```diff
@@ -784,6 +798,130 @@ public function proxyHandler(array $data, array $configuration, ?Source $proxy =
 
     }//end proxyHandler()
 
+    public function useRelayRating(array &$config): bool
+    {
+        $returnValue = true;
+        if (isset($config['query']['_federalization_relay_rating']) === true) {
+            $returnValue = $config['query']['_federalization_relay_rating'];
+
+            unset($config['query']['_federalization_relay_rating']);
+        }
+
+        return $returnValue;
+
+    }//end useRelayRating()
+
+    public function getFederationSources(array &$config, Collection $proxies): Collection
+    {
+        if (isset($config['query']['_federalization_use_sources']) === true && isset($config['query']['_federalization_exclude_sources']) === true) {
+            throw new Exception('Use of sources and exclusion of sources cannot be done in the same request');
+        }
+
+        $usedSourceIds     = [];
+        $excludedSourceIds = [];
+
+        if ((isset($config['query']['_federalization_use_sources']) === true && $config['query']['_federalization_use_sources'] === '*')
+            || (isset($config['query']['_federalization_exclude_sources']) === true && $config['query']['_federalization_exclude_sources'] === null)
+            || (isset($config['query']['_federalization_use_sources']) === false && isset($config['query']['_federalization_exclude_sources']) === false)
+        ) {
+            unset($config['query']['_federalization_exclude_sources'], $config['query']['_federalization_use_sources']);
+            return $proxies;
+        } else if (isset($config['query']['_federalization_use_sources']) === true && $config['query']['_federalization_use_sources'] !== '*') {
+            $usedSourceIds = explode(',', $config['query']['_federalization_use_sources']);
```

Comment:
What if i want to use references instead of ids here?
And could _federalization_use_sources be an array as well?

Reply:
See documentation: this is _always_ the uuid